### PR TITLE
Implement hostapi methods

### LINKED ISSUE #148: Implement HostApi methods

### PR DESCRIPTION
**Pull request by ablax**
* Separate HostApi functions
* Implement some of the HostApi functions

Closes #148

### CODE REVIEW

File: src/main/java/com/limechain/runtime/hostapi/functions/TireHostFunctions.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import lombok.experimental.UtilityClass;
+import org.wasmer.ImportObject;
+import org.wasmer.Type;
+
+import java.util.Arrays;
+import java.util.List;
+
+@UtilityClass
+public class TireHostFunctions {
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/functions/TireHostFunctions.java]:
```suggestion
public class TrieHostFunctions {
```

File: src/main/java/com/limechain/storage/DBRepository.java [hunk 2]
```diff
@@ -114,6 +117,33 @@ public synchronized boolean delete(String key) {
...
+
+                if(key.startsWith(prefix) && keysToDelete.size() < limit){
+                    keysToDelete.add(key);
+                }
+            }
+
+            for (String keyToDelete : keysToDelete) {
+                db.delete(keyToDelete.getBytes());
+            }
+            return  keysToDelete.size();
```
Comment (vikinatora) [inline src/main/java/com/limechain/storage/DBRepository.java]: double white space
```suggestion
            return keysToDelete.size();
```

File: src/main/java/com/limechain/storage/DBRepository.java [hunk 3]
```diff
@@ -114,6 +117,33 @@ public synchronized boolean delete(String key) {
...
 
+    @Override
+    public synchronized int deletePrefix(String prefixSeek, int limit) {
+        log.log(Level.INFO, String.format("deleting %d keys with prefix '%s'", limit, prefixSeek));
+        try {
+            String prefix = new String(getPrefixedKey(prefixSeek));
+            List<String> keysToDelete = new ArrayList<>();
+
+            RocksIterator rocksIterator = db.newIterator();
+            for(rocksIterator.seek(prefix.getBytes()); rocksIterator.isValid(); rocksIterator.next()){
```
Comment (vikinatora) [inline src/main/java/com/limechain/storage/DBRepository.java]: Add whitespace
```suggestion
            for (rocksIterator.seek(prefix.getBytes()); rocksIterator.isValid(); rocksIterator.next()){
```

File: src/main/java/com/limechain/runtime/hostapi/functions/OffchainHostFunctions.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                HostFunctions.getImportObject("ext_offchain_http_response_headers_version_1", argv -> {
+                    return argv;
+                }, List.of(Type.I32), Type.I64),
+                HostFunctions.getImportObject("ext_offchain_http_response_read_body_version_1", argv -> {
+                    return argv;
+                }, List.of(Type.I32, Type.I64, Type.I64), Type.I64),
+                /*
+                 * Offchain index
+                 */
+                HostFunctions.getImportObject("ext_offchain_index_set_version_1", argv -> {
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/functions/OffchainHostFunctions.java]: Is this comment really needed? There are no comments for the other functions. Either add for all of them or remove this one in order to be consistent.

File: src/main/java/com/limechain/runtime/Runtime.java [hunk 5]
```diff
@@ -32,4 +31,8 @@ public Object call(String functionName) {
         //TODO Call adequate params
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/Runtime.java:31]: This TODO isn't clear. `adequate` is has unclear meaning. Please refactor it.

File: src/main/java/com/limechain/runtime/hostapi/HostApi.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+                .allocate(8)
+                .order(ByteOrder.LITTLE_ENDIAN)
+                .putLong(res3);
+
+        return buffer.array();
+    }
+
+    public static void extAllocatorFreeVersion1(int i) {
+        //TODO: Try marking the part of the bytebuffer as free?
+    }
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/HostApi.java]: Function with empty body. Either throw unimplemented error or remove it. Also please be more specific about the comment. I don't quite understand it.

File: src/main/java/com/limechain/runtime/hostapi/HostApi.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+    public static long extStorageClearPrefixVersion2(long prefixPtr, long limitPtr) {
+        String prefix = new String(getDataFromMemory(prefixPtr));
+        int limit = new BigInteger(getDataFromMemory(limitPtr)).intValue();
+
+        int deleted = repository.deletePrefix(prefix, limit);
+        //TODO: Count how many are left?
+
+        return 0;
+    }
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/HostApi.java]: Shouldn't these be in their specific host function family class?

File: src/main/java/com/limechain/runtime/hostapi/HostApi.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        byte[] dataToHash = getDataFromMemory(addr);
+
+        byte[] hash0 = hash64(0, dataToHash);
+        byte[] hash1 = hash64(1, dataToHash);
+        byte[] hash2 = hash64(2, dataToHash);
+        byte[] hash3 = hash64(3, dataToHash);
+
+        ByteBuffer buffer = ByteBuffer.allocate(32);
+        buffer.put(hash0);
+        buffer.put(hash1);
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/HostApi.java]: Shouldn't these be in their specific host function family class?

File: src/main/java/com/limechain/runtime/hostapi/functions/HostFunctions.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.UnaryOperator;
+
+@UtilityClass
+public class HostFunctions {
+
+    static final List<Number> EMPTY_LIST_OF_NUMBER = List.of();
+    static final List<Type> EMPTY_LIST_OF_TYPES = List.of();
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/functions/HostFunctions.java]: Be specific about the access modifier to be consistent with other parts of the code.

**Review by vikinatora — CHANGES_REQUESTED** [review]
Looks good functionality wise. I'd like you to add java doc to the classes and functions and run the linter as I noticed that there are some instances of double white spaces("  ") and inconsistent formatting

File: src/main/java/com/limechain/runtime/Runtime.java (hunk 5 above)
Comment (ablax) [reply src/main/java/com/limechain/runtime/Runtime.java:31]: I'm not sure to what to refactor it. It was added from Bobi, not from me

File: src/main/java/com/limechain/runtime/hostapi/HostApi.java [hunk 10]
```diff
@@ -2,56 +2,52 @@
...
 import java.nio.ByteBuffer;
-import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.UnaryOperator;
 import java.util.logging.Level;
 
+/**
+ * Holds some common methods and services used by the different
```
Comment (vikinatora) [inline src/main/java/com/limechain/runtime/hostapi/HostApi.java]:
```suggestion
 * Holds common methods and services used by the different
```

**Review by vikinatora — APPROVED** [review]

**Pull request merged**